src/ThreadId.h

```diff
@@ -0,0 +1,75 @@
+#ifndef _THREADID_H_
+#define _THREADID_H_
+
+#include <atomic>
+#include <cstdint>
+
+namespace pipert {
+
+/// (Part of internal implementation.)
+/// Helps in tagging and identifying the currently running thread.
+///
+/// The implementation uses `thread_local` variables which are supposed to have
+/// better performing implementation than identification by using OS calls.
+/// This class is needed by the monitoring / logging system to be able to log
+/// which thread is doing what.
+/// This information is supposed to give the connection between the source
+/// and sink user _Nodes_ and connected Channel objects when data is going
+/// through a transformation done by user code in a _Node_.
+class ThreadId {
+ public:
+  /// Returns the ThreadId of the currently running thread.
+  ///
+  /// If the currently running thread was never tagged before
+  /// (see IsCurrentThreadUntagged()), it will automatically tag it
+  /// as _auto-tagged_ and return the new tag.
+  /// This way, all threads are tagged but can be later classified by
+  /// IsHandTagged() and IsAutoTagged() functions.
+  /// \return ThreadId of the currently running thread.
+  ///         The returned object cannot be untagged, but it can be either
+  ///         previously _hand tagged_ or _auto tagged_.
+  static ThreadId GetCurrentThread();
+
+  /// This explicitly tags the currently running thread as _hand tagged_,
+  /// so we can see that this is ours later on.
+  /// \pre The current thread needs to be _untagged_.
+  ///      It would lead to errors if retagging was possible.
+  static void TagCurrentThread();
+
+  /// Returns true if the current thread was not tagged or autotagged before.
+  /// See GetCurrentThread() for details on autotagging.
+  static bool IsCurrentThreadUntagged();
+
+  ThreadId(const ThreadId&) = default;
+  ThreadId& operator=(const ThreadId&) = default;
+
+  /// Returns true if the thread was tagged by explicitly calling
+  /// TagCurrentThread().
+  bool IsHandTagged() const;
```

Comment:
I am just interested, but why do we need hand tagged threads? Where will we use them in monitoring or what will be the benefits of them?

Reply:
It is basically a one bit information. The hand tagged ones will be the ones coming from the worker thread pool. So if it was autotagged, we know it is a user thread and not ours. This way, we will be identifying the entry and exit points of the system (and the user threads) as well, not just the internal connections and our own threads. It is also useful because we will be able to map up, how many user threads are there in the system (just one main thread, or several others). All is happening to be able to log these for the analyzer.